File: core/contractsdk/cpp/example/score_record.cc
#include "xchain/xchain.h"

// 学生成绩上链存证API规范
// 参数由Context提供
class ScoreRecord {
public:
    // 初始化写入权限
    // 参数: owner - 具有写入权限的address
    virtual void initialize() = 0;

    // 写入课程成绩
    // 参数: userid - 学生的主键id
    //      data - 学生的成绩信息(json格式string)
    virtual void addScore() = 0;

    // 按照学生id查询成绩
    // 参数: userid - 学生的主键id
    // 返回值: data - 学生的成绩信息(json格式string)
    virtual void queryScore() = 0;

    // 查询具有写权限的账户
    // 返回值: 具有写权限的address
    virtual void queryOwner() = 0;
};

struct ScoreRecordDemo : public ScoreRecord, public xchain::Contract {
private:
    // define the key prefix of buckets
    const std::string OWNER_KEY = "Owner";
    const std::string RECORD_KEY = "R_";

    // check if caller is the owner of this contract
    bool isOwner(xchain::Context* ctx, const std::string& caller) {
        std::string owner;
        if (!ctx->get_object(OWNER_KEY, &owner)) {
            return false;
        }
        return (owner == caller);
    }

public:
    void initialize() {
        // 获取合约上下文对象
        xchain::Context* ctx = this->context();
        // 从合约上下文中获取合约参数, 由合约部署者指定具有写入权限的address
        std::string owner = ctx->arg("owner");
        if (owner.empty()) {
            ctx->error("missing owner address");
            return;
        }
        // 将具有写入权限的owner地址记录在区块链账本中
        ctx->put_object(OWNER_KEY, owner);
        ctx->ok("success");
    }

    void addScore() {
        // 获取合约上下文对象
        xchain::Context* ctx = this->context();
        // 获取发起者身份
        const std::string& caller = ctx->initiator();
        if (caller.empty()) {
            ctx->error("missing initiator");
            return;
        }
        // 如果写操作发起者不是具有写权限的用户，则无权写入
        if (!isOwner(ctx, caller)) {
            ctx->error(
                "permission check failed, only the owner can add score record");
            return;
        }

        // 从参数中获取用户主键id，必填参数，没有则返回错误
        const std::string& userid = ctx->arg("userid");
        if (userid.empty()) {
            ctx->error("missing 'userid'");
            return;
        }
        // 从参数中获取成绩数据，必填参数，没有则返回错误
        const std::string& data = ctx->arg("data");
        if (data.empty()) {
            ctx->error("missing 'data'");
            return;
        }

        // 将具有写入权限的owner地址记录在区块链账本中
        std::string score_key = RECORD_KEY + userid;
        if (!ctx->put_object(score_key, data)) {
            ctx->error("failed to save score record");
            return;
        }

        // 执行成功，返回status code 200
        ctx->ok(userid);
    }

    void queryScore() {
        // 获取合约上下文对象
        xchain::Context* ctx = this->context();

        // 从参数中获取用户主键id，必填参数，没有则返回错误
        const std::string& userid = ctx->arg("userid");
        if (userid.empty()) {
            ctx->error("missing 'userid'");
            return;
        }

        // 从账本中读取学生的成绩数据
        std::string score_key = RECORD_KEY + userid;
        std::string data;
        if (!ctx->get_object(score_key, &data)) {
            // 没查到，说明之前没上链过，返回错误
            ctx->error("no score record found of " + userid);
            return;
        }

        // 执行成功，返回status code 200
        ctx->ok(data);
    }

    void queryOwner() {
        // 获取合约上下文对象
        xchain::Context* ctx = this->context();
        std::string owner;
        if (!ctx->get_object(OWNER_KEY, &owner)) {
            // 没查到owner信息，可能
            ctx->error("get owner failed");
            return;
        }
        // 执行成功，返回owner address
        ctx->ok(owner);
    }
};

DEFINE_METHOD(ScoreRecordDemo, initialize) { self.initialize(); }

DEFINE_METHOD(ScoreRecordDemo, addScore) { self.addScore(); }

DEFINE_METHOD(ScoreRecordDemo, queryScore) { self.queryScore(); }

DEFINE_METHOD(ScoreRecordDemo, queryOwner) { self.queryOwner(); }

File: core/contractsdk/java/example/score_record/src/main/java/com/baidu/xuper/ScoreRecord.java
package com.baidu.xuper;


/**
 * Counter
 */
public class ScoreRecord implements Contract {
    private final String OWNER_KEY = "owner";
    private final String RECORD_KEY = "R_";

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        byte[] owner = ctx.args().get("owner");
        if (owner == null || owner.length == 0) {
            return Response.error("missing caller");
        }
        ctx.putObject(OWNER_KEY.getBytes(), owner);
        return Response.ok("ok~".getBytes());
    }

    @ContractMethod
    public Response addScore(Context ctx) {
        String caller = ctx.caller();
        if (caller == null || caller.isEmpty()) {
            return Response.error("missing caller");
        }
        byte[] userIdByte = ctx.args().get("user_id");
        byte[] dataByte = ctx.args().get("data");
        if (userIdByte == null || userIdByte.length == 0) {
            return Response.error("missing user_id");
        }
        if (dataByte == null || dataByte.length == 0) {
            return Response.error("missing data");
        }
        if (!new String(ctx.getObject(OWNER_KEY.getBytes())).equals(caller)) {
            return Response.error("you do not have permission to call this method");
        }
        ctx.putObject((RECORD_KEY + new String(userIdByte)).getBytes(), dataByte);
        return Response.ok(userIdByte);
    }

    @ContractMethod
    public Response queryScore(Context ctx) {
        byte[] userIdByte = ctx.args().get("user_id");
        if (userIdByte == null || userIdByte.length == 0) {
            return Response.error("missing user_id");
        }
        byte[] data = ctx.getObject((RECORD_KEY + new String(userIdByte)).getBytes());
        if (data == null) {
            return Response.error("record of " + new String(userIdByte) + " not found'");
        }
        return Response.ok(data);
    }

    @ContractMethod
    public Response queryOwner(Context ctx) {
        return Response.ok(ctx.getObject(OWNER_KEY.getBytes()));
    }

    public static void main(String[] args) {
        Driver.serve(new ScoreRecord());
    }
}
